junebug/channel.py

```diff
@@ -0,0 +1,149 @@
+}
+
+
+class Channel(object):
+    def __init__(
+            self, redis_manager, amqp_config, properties, id=None):
+        '''Creates a new channel. ``redis_manager`` is the redis manager, from
+        which a sub manager is created using the channel id. If the channel id
+        is not supplied, a UUID one is generated. Call ``save`` to save the
+        channel data. It can be started using the ``start`` function.'''
+        self._properties, self.id, self.redis = (
+            properties, id, redis_manager)
+        if self.id is None:
+            self.id = str(uuid.uuid4())
+
+        self.options = deepcopy(VumiOptions.default_vumi_options)
+        self.options.update(amqp_config)
+
+    def _convert_unicode(self, data):
+        # Twisted doesn't like it when we give unicode in for config things
+        if isinstance(data, basestring):
+            return str(data)
+        elif isinstance(data, collections.Mapping):
+            return dict(map(self._convert_unicode, data.iteritems()))
+        elif isinstance(data, collections.Iterable):
+            return type(data)(map(self._convert_unicode, data))
+        else:
+            return data
+
+    @inlineCallbacks
+    def start(self, service, transport_worker=None):
+        '''Starts the relevant workers for the channel. ``service`` is the
+        parent of under which the workers should be started.'''
+        class_name = transports.get(self._properties.get('type'))
+        if class_name is None:
+            raise InvalidChannelType(
+                'Invalid channel type %r, must be one of: %s' % (
+                    self._properties.get('type'),
+                    ', '.join(transports.keys())))
+
+        # transport_worker parameter is for testing, if it isn't specified,
+        # create the transport worker
+        if transport_worker is None:
+            workercreator = WorkerCreator(self.options)
+            config = self._convert_unicode(self._properties['config'])
+            transport_worker = workercreator.create_worker(
+                class_name, config)
+        transport_worker.setName(self.id)
+        transport_worker.setServiceParent(service)
+        yield transport_worker.startService()
+        self.transport_worker = transport_worker
+
+    @inlineCallbacks
+    def stop(self):
+        '''Stops the relevant workers for the channel'''
+        if hasattr(self, 'transport_worker'):
+            yield self.transport_worker.stopService()
+            yield self.transport_worker.disownServiceParent()
+            del self.transport_worker
+
+    @inlineCallbacks
+    def save(self):
+        '''Saves the channel data into redis.'''
+        properties = json.dumps(self._properties)
+        channel_redis = yield self.redis.sub_manager(self.id)
+        yield channel_redis.set('properties', properties)
+        yield self.redis.sadd('channels', self.id)
+        yield channel_redis.close_manager()
```

Comment:
This is one thing that I'm not sure about with the new way of doing this. Do we need to close the `sub_manager` here? Or does it get closed when we close the root manager?